pkg/operator/cluster/creds_controller.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2016 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Some of the code below came from https://github.com/coreos/etcd-operator
+which also has the apache 2.0 license.
+*/
+
+package cluster
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/api/core/v1"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+var (
+	// controllerResyncPeriod potentially can be part of rook Operator configuration.
+	// in seconds
+	controllerResyncPeriod time.Duration = 120
+	// controllerRetry is number of retries to process an event
+	controllerRetry int = 5
+)
+
+func NewCredsController(ns string, context *clusterd.Context) *PVCController {
+	return &PVCController{
+		clusterNamespace: ns,
+		clusterContext:   context,
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+}
+
+// PVCController implements Kube controller that monitors rook PVC events.
+// The controller ensures all kube namespaces have a set of credentials
+// to manage lifeCycle of ceph primitives.
+type PVCController struct {
+	queue            workqueue.RateLimitingInterface
+	indexer          cache.Indexer
+	informer         cache.Controller
+	clusterContext   *clusterd.Context
+	clusterNamespace string
+}
+
+func (c *PVCController) pvcHandler(key interface{}) error {
+	obj, _, err := c.indexer.GetByKey(key.(string))
+	if err != nil {
+		return err
+	}
+
+	pvc := obj.(*v1.PersistentVolumeClaim)
+
+	if !strings.Contains(*pvc.Spec.StorageClassName, "rook") {
+		return nil
+	}
+
+	cephUser := &cephUser{c.clusterContext, c.clusterNamespace}
+	if err := cephUser.create(); err != nil {
+		return err
+	}
+
+	if err = cephUser.setKubeSecret(pvc.ObjectMeta.Namespace); err != nil {
```

Comment:
What is the timing of processing this event compared to when the PVC is created? Is there a race condition? If we create the secret asynchronously, the secret might not be created before it is needed by the PVC. How can we guarantee the secret is created in time?